### DISCUSSION TITLE
Clarification on "set" task output

### DISCUSSION THREAD

**Discussion by alvin** [serverlessworkflow/specification#1015] [Q&A] [answered]
Hello, I'm working on a small sub-set implementation of the specification for my use-case.

I've spent some time pouring over the details and I can't tell if this test case is something I should be aiming for or not.  Should a "set" task's mutations be the output of a task or should they mutate $context or ?

**Basically -- does this scenario express something true about the spec or not?** 
```

    it("should not output context values as task $output without explicit output configuration", async () => {
      const workflow = {
        ...commonWorkflow,
        do: [
          {
            setValues: {
              set: {
                result: {
                  doubledNumber: "${ $workflow.input.number * 2 }",
                },
              },
            },
          },
        ],
        // No output configuration, so the set values should not be in the result
      };

      const result = await executeWorkflow(workflow, commonInput);
      expect(result).toEqual({});
    });
```

Overall I'm a bit uncertain about the role of $context vs. the $output cascade... I've made some progress in understanding over the past few days, but it's a work in progress :)

**Comment by cdavernas** [comment, marked as answer]
> Should a "set" task's mutations be the output of a task or should they mutate $context or ?

The `set` task mutates its own input. If you do not configure output, its mutated input will become the output of the task.

> Basically -- does this scenario express something true about the spec or not?

No, the assumption made in your scenario, as far I understand it, is wrong. 

As a matter of fact, the result of the `setValues` task, without output configuration, should be (given its input was `{}` and the `$workflow.input.number` was, say, `3`):

```yaml
result:
  doubledNumber: 6
```

Because you did not configure output at flow level either, the output of the workflow will be the above, too.

> Overall I'm a bit uncertain about the role of $context vs. the $output cascade... 

You can see `$context` as a value bag that you can use throughout your flow, but which has no implicit on any output. Output is used to set the "final" data returned by a task or a flow.

I hope this answers your questions!

**Reply by alvin** [reply]
Thank you so much, yes I understood this fundamental distinction between $context and $output, but in the context of "set" directives it somehow felt really tricky to disambiguate!  I really appreciate the help :)

**Reply by cdavernas** [reply]
Come on, very glad I could be of assistance!
